src/app/frontend/deploy/deploy_label_service.js

```diff
@@ -0,0 +1,160 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+import DeployFormLabel, {APP_LABEL_KEY, VERSION_LABEL_KEY} from './deploy_form_label';
+
+/**
+ * Service used for handling label actions like: hover, showing duplicated key error, etc.
+ * @final DeployLabelService
+ */
+export default class DeployLabelService {
+
+  /**
+   * Checks label and either adds/deletes row to/from list if there is no error.
+   * @param {!Array<!DeployFormLabel>} labelList
+   * @param {!DeployFormLabel} label
+   */
+  checkLabel(labelList, label) {
+    /** @type {!DeployFormLabel} */
+    let lastLabel = labelList[labelList.length - 1];
+
+    if (label.key === '' && label.value === '' && label !== lastLabel) {
+      this.deleteLabel(labelList, label);
+    }
+
+    if (lastLabel.key !== '' && lastLabel.value !== '') {
+      this.addLabel(labelList);
+    }
+  }
+
+  /**
+   * Adds row to labels list.
+   * @param {!Array<!DeployFormLabel>} labelList
+   */
+  addLabel(labelList) {
+    labelList.push(new DeployFormLabel());
+  }
+
+  /**
+   * Deletes row from labels list.
+   * @param {!Array<!DeployFormLabel>} labelList
+   * @param {!DeployFormLabel} label
+   */
+  deleteLabel(labelList, label) {
+    for (let i = 0; i < labelList.length; i++) {
```

Comment:
Use Array.prototype.findIndex and then remove the element from the index. See http://www.ecma-international.org/ecma-262/6.0/#sec-properties-of-the-array-prototype-object

Such code is less error prone, because you do not have iteration.